cql3/restrictions/restriction.hh

```diff
@@ -58,6 +61,52 @@ namespace restrictions {
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+/// Work-in-progress new restriction representation.  When complete, it will replace the old one.
+namespace wip {
+
+// The new representation exposes its member data publicly.  Operations on restrictions are
+// performed by free functions that take restrictions as parameters and use the visitor pattern to
+// specialize code for different kinds of restrictions.
+//
+// The most interesting class is wip::oper below, which can represent both multi- and single-column
+// restrictions.  Instead of the old merge_with mechanism for combining restrictions, we will simply
+// add them into a conjunction expression that will be processed using visitors.
+
+/// Represents token function on LHS of an operator relation.  No need to list column definitions
+/// here -- token takes exactly the partition key as its argument.
+struct token {};
+
+/// Identifier(s) on LHS of an operator restriction.
+using idents = std::vector<const column_definition*>;
+
+/// Subscripted column(s) on LHS of an operator restriction.  Eg: `m[123] > 456`.  Able to represent
+/// multiple columns, eg, `(x[1], y[2]) = (3, 4)`.
+struct subscript {
+    idents columns;
+    std::vector<::shared_ptr<term>> subs; ///< Corresponding subscripts.
+};
+
+/// LHS of an operator restriction.
+using lhs_t = std::variant<idents, token, subscript>;
+
+/// Base class for all restrictions.
+struct node {
+    virtual ~node() = default;
+};
+
```

Comment:
Why not std::variant<all different node types>? That makes it easy to concentrate operations over nodes in one place, rather than spreading them out over the code base in derived classes and virtual function overloads as is the case now.

Reply:
There will be no virtual function overloads, as there will be no methods -- only free functions using visitors, concentrating logic in one place.  This can be accomplished either via `variant` or via subtyping (ie, inheritance).  I'm trying out both before I decide which I prefer for this effort, and then I'll use that across the board.  (Which means this may very well become `variant`.)

The drawbacks of subtyping are
1. You typically need a pre-build procedure for generating the visitor base class from the inheritance hierarchy.
2. The most common implementation choices are double-dispatch through virtual tables or heavily templated, brittle code, neither of which is very satisfying.

The drawbacks of `variant` are
1. Specific types must be defined before general ones, hampering readability.  This is why, for instance, I have to define subtypes of `lhs_t` before `lhs_t` itself.  It makes blurbs awkward and prompts the reader to constantly jump down the file to the more general explanations.
2. Visitors for deep hierarchies can be awkward to write.  If, for instance, we have this:
```bnf
restriction ::= primary | '(' primary [ ',' primary ]* ')'
primary ::= predicate | boolean_cast
predicate ::= comparison | contains_predicate | like_predicate | in_predicate
```
and we need a visitor that finds all like_predicates in a restriction, that's easy to write with subtyping but not with `variant`.

Reply:
```c++
    std::vector<primary> restrictions;
    auto has_like_predicate = boost::any(restrictions, overloaded(
        [] (predicate& p) {
             return overloaded(
                  [] (like_predicate& lp) { return true; }
                  [] (auto& default_case) { return false; } 
             });
        [] (auto& default_case) { return false; }
    });
```

so while it's not great, it's also not terrible, if you have `overloaded` and the default_case catch-alls (trick from @espindola).

Reply:
Oh, we definitely have `utils/overloaded_functor.hh`. :)  But compare the above to LLVM's subtyping-based [visitor](http://llvm.org/doxygen/InstVisitor_8h_source.html#l00045).  And then multiply the extra boilerplate by the total number of visitors implemented...

I don't know yet how deep our hierarchy will get; right now it's quite shallow, and `variant` visitors are quite acceptable.

Reply:
Also, you're cheating above by using `vector` -- that should be an extra layer of depth `[] (restriction& r) { ... }`

Reply:
First, I reserve the right to cheat.

Second, if you wrote `std::variant<primary, std::vector<primary>>` you'd get a review comment asking you to simplify to std::vector. 

Third, I acknowledge that it's not perfect. For deeper hierarchies we'd split the nested thing above into multiple free functions. Still I think it's better than subtyping and virtuals, since you don't need to touch every source file to check if an AST has a like_matcher somewhere.

Reply:
Fourth, we might find a way for `variant` visitors to reuse generic boilerplate with just a line of code...

Reply:
> Still I think it's better than subtyping and virtuals, since you don't need to touch every source file to check if an AST has a like_matcher somewhere.

Why would we need "to touch every source file" with subtyping??

Reply:
subtyping+virtuals. Every class has to implement has_like_matcher().

Strictly speaking we can prune the tree at the top by using defaults. But that defaults are dangeous, they embed knowledge that was true at the time the default was written, and can be not updated when a new subtype is added. The supertype should not encode knowledge about its subtypes, and when you add a default, it does.

Reply:
> subtyping+virtuals. Every class has to implement has_like_matcher().

Um, no.  That's not how a subtyping visitor works...  Perhaps I wasn't clear in all my previous comments and links?

> Strictly speaking we can prune the tree at the top by using defaults. But that defaults are dangeous, they embed knowledge that was true at the time the default was written, and can be not updated when a new subtype is added.

This is why most subtyping visitor implementations have a pre-build procedure I mentioned upthread.

Reply:
BTW, to prevent `vector` cheating in the code above, I hereby amend the first production to a more realistic version:
```
restriction ::= primary | primary 'AND' restriction | primary 'OR' restriction | '(' restriction ')'
```
:smiley:

Reply:
I'd rather avoid code generation. I was referring to non-visitor patterns like we have now. If you refer to the standard visitor pattern with accept/visit virtual functions, I think it's strictly inferior to std::variant.

Reply:
```c++
    bool has_like_matcher(primary& p) { /* the overloaded thing */ }
    bool has_like_matcher(predicate& p) { ... }
    bool has_like_matcher(primary& p) { ... }
    bool has_like_matcher(binary_boolean_operation& bbo) { ... }
    bool has_like_matcher(restriction& r) { ... }

    restriction restrictions;
    return has_like_matcher(restrictions);
```

It takes some work, but nothing like code generation.

Reply:
> I'd rather avoid code generation.

In my experience, it works OK.  I wouldn't hesitate to rely on it in case of a sufficiently deep subtype hierarchy.

> I was referring to non-visitor patterns like we have now. If you refer to the standard visitor pattern

Both that and other subtyping visitors, such as LLVM's.  Even dynamic_cast comes under this heading; I used it in this RFC as a cheap draft of what the visitor code would look like.

> with accept/visit virtual functions, I think it's strictly inferior to std::variant.

The "strictly" part overstates the case.  More readability and less boilerplate are the opposite of inferior.

Reply:
Replying to @espindola's comment in-thread:
> I think the more common pattern in LLVM is the one documented here: https://www.llvm.org/docs/HowToSetUpLLVMStyleRTTI.html 

True, though that writeup doesn't supply a nifty example illustrating boilerplate absence in the use-case we're discussing.  Since it's just a faster dynamic_cast, it falls into the same broad category of subtyping-based "visitors".  Only instead of `visitFoo(Foo& f) {...}` you write `if (auto f = dyn_cast<Foo>(x)) {...}`.